server/src/actors/events.rs

```diff
@@ -125,6 +142,162 @@ impl EventProcessor {
         Self { rate_limiter }
     }
 
+    fn message_pack_to_annotated_breadcrumbs(item: &Item) -> Option<Array<Breadcrumb>> {
+        let payload = item.payload();
+        let des = &mut mps::Deserializer::from_slice(&payload);
+
+        Annotated::<Array<Breadcrumb>>::deserialize_with_meta(des)
+            .ok()
+            .and_then(Annotated::into_value)
+    }
+
+    fn add_message_pack_breadcrumbs(
+        evt: &mut Annotated<Event>,
+        breadcrumbs1: Option<Item>,
+        breadcrumbs2: Option<Item>,
+    ) {
+        let mut breadcrumbs1 = breadcrumbs1
+            .as_ref()
+            .and_then(Self::message_pack_to_annotated_breadcrumbs)
+            .unwrap_or_default();
+
+        let mut breadcrumbs2 = breadcrumbs2
+            .as_ref()
+            .and_then(Self::message_pack_to_annotated_breadcrumbs)
+            .unwrap_or_default();
+
+        let timestamp1 = breadcrumbs1
+            .iter()
+            .rev()
+            .find_map(|breadcrumb| breadcrumb.value().and_then(|b| b.timestamp.value()));
+
+        let timestamp2 = breadcrumbs2
+            .iter()
+            .rev()
+            .find_map(|breadcrumb| breadcrumb.value().and_then(|b| b.timestamp.value()));
+
+        if timestamp1 > timestamp2 {
+            std::mem::swap(&mut breadcrumbs1, &mut breadcrumbs2);
+        }
+
+        let max_length = std::cmp::max(breadcrumbs1.len(), breadcrumbs2.len());
+
+        breadcrumbs1.extend(breadcrumbs2);
+
+        if breadcrumbs1.len() > max_length {
+            // keep only the last max_length elements from the vectors
+            breadcrumbs1.drain(0..(breadcrumbs1.len() - max_length));
+        }
+
+        if let Some(evt) = evt.value_mut() {
+            evt.breadcrumbs = Annotated::new(Values {
+                values: Annotated::new(breadcrumbs1.into()),
+                other: Object::default(),
+            });
+        }
+    }
+
+    /// Writes a placeholder to indicate that this event has an associated minidump.
+    fn write_minidump_placeholder(evt: &mut Annotated<Event>) {
+        let evt = evt.get_or_insert_with(Event::default);
+
+        let platform = evt.platform.value_mut();
+        *platform = Some("native".to_string());
+
+        let level = evt.level.value_mut();
+        *level = Some(Level::Fatal);
+
+        let exceptions = evt
+            .exceptions
+            .value_mut()
+            .get_or_insert_with(Values::<Exception>::default)
```

Comment:
The type can be inferred by the compiler. 

```suggestion
            .get_or_insert_with(Values::default)
```

On a side note, would it be more convenient for us to add an `Annotated::get_or_default` method? It looks like we're doing this a lot across the codebase now. No need to add this now, though.